opentracing/ext/scope_manager/tornado.py

```diff
@@ -0,0 +1,280 @@
+
+            @tornado.gen.coroutine
+            def coroutine(input):
+                # No span should be activated here.
+                # The parent Span will remain active, though.
+                with tracer.start_span('child', child_of=tracer.active_span):
+                    ...
+
+            @tornado.gen.coroutine
+            def handle_request_wrapper():
+                res1 = corotuine('A')
+                res2 = corotuine('B')
+
+                yield [res1, res2]
+    """
+
+    def activate(self, span, finish_on_close):
+        """
+        Make a :class:`~opentracing.Span` instance active.
+
+        :param span: the :class:`~opentracing.Span` that should become active.
+        :param finish_on_close: whether *span* should automatically be
+            finished when :meth:`Scope.close()` is called.
+
+        If no :func:`tracer_stack_context()` is detected, thread-local
+        storage will be used to store the :class:`~opentracing.Scope`.
+        Observe that in this case the active :class:`~opentracing.Span`
+        will not be automatically propagated to the child corotuines.
+
+        :return: a :class:`~opentracing.Scope` instance to control the end
+            of the active period for the :class:`~opentracing.Span`.
+            It is a programming error to neglect to call :meth:`Scope.close()`
+            on the returned instance.
+        """
+
+        context = self._get_context()
+        if context is None:
+            return super(TornadoScopeManager, self).activate(span,
+                                                             finish_on_close)
+
+        scope = _TornadoScope(self, span, finish_on_close)
+        context.active = scope
+
+        return scope
+
+    @property
+    def active(self):
+        """
+        Return the currently active :class:`~opentracing.Scope` which
+        can be used to access the currently active
+        :attr:`Scope.span`.
+
+        :return: the :class:`~opentracing.Scope` that is active,
+            or ``None`` if not available.
+        """
+
+        context = self._get_context()
+        if not context:
+            return super(TornadoScopeManager, self).active
+
+        return context.active
+
+    def _get_context(self):
+        return _TracerRequestContextManager.current_context()
+
+
+class _TornadoScope(Scope):
+    def __init__(self, manager, span, finish_on_close):
+        super(_TornadoScope, self).__init__(manager, span)
+        self._finish_on_close = finish_on_close
+        self._to_restore = manager.active
+
+    def close(self):
+        context = self.manager._get_context()
+        if context is None or context.active is not self:
+            return
+
+        context.active = self._to_restore
+
+        if self._finish_on_close:
+            self.span.finish()
+
+
+class ThreadSafeStackContext(tornado.stack_context.StackContext):
+    """
+    Thread safe version of Tornado's StackContext (up to 4.3)
```

Comment:
We're using a similar implementation with good enough results. But I'll leave to @yurishkuro a deeper review.

Reply:
The Tornado core maintainer wrote a message about removing the `StackContext` starting from Tornado 6.0 ([ref](https://github.com/tornadoweb/tornado/issues/1063)). The new `contextvars` module in [Python 3.7](https://docs.python.org/3.7/whatsnew/3.7.html#whatsnew37-pep567) will be the official way to handle the Context.

This is something we need to support to provide a better support when Python 3.7 will be used. Let's keep it in our radar for the future. cc @carlosalberto